internal/pkg/validation/state.go

```diff
@@ -0,0 +1,295 @@
+package validation
+
+import (
+	"context"
+	"fmt"
+	"github.com/pkg/errors"
+	"math/rand"
+
+	vstate "github.com/filecoin-project/chain-validation/pkg/state"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-hamt-ipld"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+
+	"github.com/filecoin-project/go-filecoin/internal/pkg/crypto"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/types"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/actor"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/actor/builtin/initactor"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/actor/builtin/storagemarket"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/address"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/state"
+)
+
+// StateWrapper provides a wrapper for a state tree, storage, map and keystore.
+type StateWrapper struct {
+	state.Tree
+	vm.StorageMap
+	keys *keyStore
+}
+
+var _ vstate.Wrapper = &StateWrapper{}
+
+// NewState returns a StateWrapper.
+func NewState() *StateWrapper {
+	bs := blockstore.NewBlockstore(datastore.NewMapDatastore())
+	cst := hamt.CSTFromBstore(bs)
+	treeImpl := state.NewTree(cst)
+	storageImpl := vm.NewStorageMap(bs)
+	return &StateWrapper{treeImpl, storageImpl, newKeyStore()}
+}
+
+// Cid returns the cid of the state wrappers current state
+func (s *StateWrapper) Cid() cid.Cid {
+	panic("implement me")
```

Comment:
Blocking: I think we should replace the state.Tree with a pattern similar to that used by the Lotus adapter, storing the state root CID and hamt.CborIpldStore in the state wrapper, and then loading the state on demand when used.